lib/ex_aws/credentials_ini/file.ex

```diff
@@ -1,15 +1,126 @@
+           {_, {:ok, %{"expiresAt" => expires_at} = content}} <-
+             {:decode, Jason.decode(content_raw)},
+           {_, :ok} <-
+             {:expiration, check_sso_expiration(expires_at)},
+           {_, {:ok, sso_creds}} <-
+             {:sso_creds, request_sso_role_credentials(content, sso_account_id, sso_role_name)},
+           {_, {:ok, reformatted_creds}} <-
+             {:rename, rename_sso_credential_keys(sso_creds)} do
+        {:ok, reformatted_creds}
+      else
+        {:read, error} -> {:error, "Could not read SSO cache file"}
+        {:decode, error} -> {:error, "SSO cache file contains invalid json"}
+        {:expiration, error} -> error
+        {:sso_creds, error} -> error
+        {:rename, error} -> error
+      end
+    end
+
+    defp get_sso_cache_file(sso_start_url) do
+      hash = :crypto.hash(:sha, sso_start_url) |> Base.encode16() |> String.downcase()
+
+      System.user_home()
+      |> Path.join(".aws/sso/cache/#{hash}.json")
+    end
+
+    defp check_sso_expiration(expires_at_str) do
+      with {:ok, expires_at, _} <- DateTime.from_iso8601(expires_at_str) do
+        case DateTime.compare(expires_at, DateTime.utc_now()) do
+          :gt -> :ok
+          _ -> {:error, "SSO access token is expired, refresh the token with `aws sso login`"}
+        end
+      else
+        {:error, err} -> {:error, "SSO cache file has invalid expiration format: #{err}"}
+      end
+    end
+
+    # TODO: Implement check that verifies if a user has aws_sso and source profile configured
+    # Saw this concern raised in saml2aws using aws go sdk and think it could be helpful
+
+    # TODO: the :ex_aws.request() abstraction that can choose hackney or what users set requires signing
+    # We can't sign yet, so this is written to hackney, but need to revisit alternatives
+    defp request_sso_role_credentials(
+           %{"accessToken" => access_token, "region" => region},
+           account_id,
+           role_name
+         ) do
+      with {_, {:ok, %{status_code: 200, headers: _headers, body: body_raw}}} <-
+             {:request,
+              request(
+                :get,
+                "https://portal.sso.#{region}.amazonaws.com/federation/credentials?account_id=#{
+                  account_id
+                }&role_name=#{role_name}",
+                "",
+                [{"x-amz-sso_bearer_token", access_token}]
+              )},
+           {_, {:ok, body}} <- {:decode, Jason.decode(body_raw)} do
+        {:ok, body}
+      else
+        {:request, {_, %{status_code: status_code} = resp}} ->
+          {:error, "SSO role credentials request responded with #{status_code}"}
+
+        {:decode, err} ->
+          {:error, "Could not decode SSO role credentials response"}
+      end
+    end
+
+    defp rename_sso_credential_keys(%{"roleCredentials" => role_credentials}) do
+      with {_, access_key} when not is_nil(access_key) <-
```

Comment:
For my own learning what brought you to this strategy? Seems very guarded for possible errors and really useful, especially in `get_sso_role_credentials` which was a very good use of `with` IMO.

For this function, I also am kind of curious do we need to be as guarded since we already error-checked the API response.

Again more just trying to keep up and learn as we are going through this not critiques :D